### PR TITLE
Calculate DOF permutations for shared dofs on edges/facets

### PR DESCRIPTION
**Pull request by chrisrichardson**
Adds a new generated function to `ufc.h` to permute dofs on cell facets/edges so that higher order elements are consistent with neighbours, now that the mesh is no longer ordered.

There still remains work to do:
* quad/hex cells (quads should be working, but hard to test until we get `evaluate_reference_basis()`
* Hdiv/Hcurl elements need extra attention

### CODE REVIEW

File: ffc/backends/ufc/dofmap.py [hunk 1]
```diff
@@ -122,6 +122,118 @@ def tabulate_facet_dofs(L, ir):
...
+    facet_edge_vertices = {'tetrahedron': (((2, 3), (1, 3), (1, 2)),
+                                           ((2, 3), (0, 3), (0, 2)),
+                                           ((1, 3), (0, 3), (0, 1)),
+                                           ((1, 2), (0, 2), (0, 1))),
+                           'hexahedron': (((0, 1), (2, 3), (0, 2), (1, 3)),
+                                          ((4, 5), (6, 7), (4, 6), (5, 7)),
+                                          ((0, 1), (4, 5), (0, 4), (1, 5)),
+                                          ((2, 3), (6, 7), (2, 6), (3, 7)),
+                                          ((0, 2), (4, 6), (0, 4), (2, 6)),
+                                          ((1, 3), (5, 7), (1, 5), (3, 7)))}
```
Comment (blechta) [inline ffc/backends/ufc/dofmap.py]: I suggest to query these table from FIAT. Here is described how to do it: https://bitbucket.org/fenics-project/ffc/issues/169/replace-tables-from-ufc_geometryh-by

File: ffc/representation.py [hunk 2]
```diff
@@ -192,6 +192,52 @@ def _compute_element_ir(ufl_element, element_numbers, classnames, parameters, ji
     return ir
 
 
+def _compute_dofmap_permutation_tables(fiat_element, cell):
+
+    if isinstance(fiat_element, MixedElement):
+        elems = fiat_element.elements()
```
Comment (blechta) [inline ffc/representation.py]: Is this correct? What about if a subelement is again a mixed element? What if it is other kind of compound element?

File: ffc/backends/ufc/dofmap_template.py [hunk 3]
```diff
@@ -18,6 +18,11 @@
 {tabulate_dofs}
 }}
 
+void tabulate_dof_permutations_{factory_name}(const int64_t* restrict global_indices, int* restrict perm, int ndofs)
```
Comment (blechta) [inline ffc/backends/ufc/dofmap_template.py]: Should we keep following the convention that the output argument is first?

File: ffc/backends/ufc/dofmap.py (hunk 1 above)
Comment (chrisrichardson) [reply ffc/backends/ufc/dofmap.py]: Thanks - I've now added that

File: ffc/representation.py (hunk 2 above)
Comment (chrisrichardson) [reply ffc/representation.py]: I don't know. I don't have any examples at the moment, so I guess we can deal with it, when it comes up.

File: ffc/backends/ufc/dofmap_template.py (hunk 3 above)
Comment (chrisrichardson) [reply ffc/backends/ufc/dofmap_template.py]: Sure. I've now reordered the args. Thanks.

**Comment by chrisrichardson** [comment]
@blechta - I've tried to add a few more descriptive comments. Maybe that helps. 
The basic idea is to generate a permutation vector, based on the orientation of the cell. For edges which are not in the reference orientation, the dofs on that edge are reversed. For facets, it is more complicated, but a similar idea.

File: ffc/backends/ufc/dofmap.py [hunk 4]
```diff
@@ -122,6 +122,111 @@ def tabulate_facet_dofs(L, ir):
...
+    code = [L.ForRange(i, 0, ndofs, body=[L.Assign(perm[i], i)])]
+
+    if tdim == 1 or (len(edge_perms) == 0 and len(facet_perms) == 0):
+        code += [L.Return()]
+        return L.StatementList(code)
+
+    global_indices = L.Symbol("global_indices")
+
+    # Collect up some topological tables
+    from FIAT.reference_element import ufc_cell
```
Comment (garth-wells) [inline ffc/backends/ufc/dofmap.py]: Can this import be at the top, and be `import FIAT.reference_element`?

File: ffc/backends/ufc/dofmap.py [hunk 5]
```diff
@@ -122,6 +122,111 @@ def tabulate_facet_dofs(L, ir):
         return L.NoOp()
 
 
+def tabulate_dof_permutations(L, ir):
```
Comment (garth-wells) [inline ffc/backends/ufc/dofmap.py]: Could this function more clearly separate (1) the concepts and (2) the code formatting (the `L` parts).

File: ffc/backends/ufc/ufc.h [hunk 6]
```diff
@@ -165,6 +165,9 @@ extern "C"
                           const int64_t* restrict num_global_entities,
                           const int64_t** entity_indices);
 
+    /// Calculate dof permutation for given edge ordering
```
Comment (garth-wells) [inline ffc/backends/ufc/ufc.h]: Could this comment be made less succinct?

File: ffc/backends/ufc/ufc.h [hunk 7]
```diff
@@ -165,6 +165,9 @@ extern "C"
...
+    void (*tabulate_dof_permutations)(int* restrict perm, const int64_t* restrict global_indices, int ndofs);
```
Comment (garth-wells) [inline ffc/backends/ufc/ufc.h]: Explain arguments. What is `ndofs`?

File: ffc/fiatinterface.py [hunk 8]
```diff
@@ -292,3 +292,22 @@ def _create_restricted_element(ufl_element):
         return MixedElement(elements)
 
     raise FFCError("Cannot create restricted element from: {}".format(ufl_element))
+
+
+def triangle_permutation_table(n, interior=0):
+
```
Comment (garth-wells) [inline ffc/fiatinterface.py]: docstring?

File: ffc/representation.py [hunk 9]
```diff
@@ -26,7 +26,7 @@
 from ffc import FFCError
 from ffc import classname
 from ffc.fiatinterface import (EnrichedElement, MixedElement, QuadratureElement, SpaceOfReals,
-                               create_element)
+                               create_element, triangle_permutation_table)
```
Comment (garth-wells) [inline ffc/representation.py]: `from ffc import fiatinterface`. Let's not perpetuate past sloppy style . . .

File: ffc/representation.py [hunk 10]
```diff
@@ -192,6 +192,54 @@ def _compute_element_ir(ufl_element, element_numbers, classnames, parameters, ji
...
+        elems = (fiat_element, )
+
+    td = cell.topological_dimension()
+
+    # Lists of permutations for each edge or facet (if any)
+    edge_permutations = []
+    face_permutations = []
+
+    offset = 0
+    for el in elems:
```
Comment (garth-wells) [inline ffc/representation.py]: Can we use more readable variable names? Could this be `for element in elements:`?

File: ffc/representation.py [hunk 11]
```diff
@@ -192,6 +192,54 @@ def _compute_element_ir(ufl_element, element_numbers, classnames, parameters, ji
...
+                for i in range(len(v)):
+                    idx1 = v[i]
+                    idx2 = v[-i - 1]
+                    if idx1 != idx2:
+                        edge_permutations[k][idx1 + offset] = idx2 + offset
+        if td > 2 and nd[2] > 1 and cell.cellname() == 'tetrahedron':
+            # Permutation on a triangular facet
+            # FIXME: add support for quadrilateral facets
+            # FIXME: add support for Hdiv/Hcurl elements
+            n = el.degree()
```
Comment (garth-wells) [inline ffc/representation.py]: `n` is a cryptic variable name in this context. Something better?

**Pull request merged**